data/cache/cache_key.py

```diff
@@ -9,6 +9,14 @@ class CacheKey(namedtuple("CacheKey", ["key", "expiration"])):
     pass
 
 
+def for_upstream_registry_token(registry, repo_name, expires_in):
+    """
+    Returns a cache key for an upstream registry auth token.
+    """
+    key = f"upstream_token__{registry}_{repo_name}"
+    return CacheKey(key, expires_in)
```

Comment:
Probably not a good idea to store the registry token. Any reason why you couldn't just pass this along the client in the response normally, just like any normal registry auth requests?

Reply:
That's not how basic auth works[1]. The challenge[2] Quay returns on 401 responses is specific to Quay. Any third-party services Quay may talk to are not part of the equation.

A forward proxy behaves that way [3][4][5], but it isn't our goal to turn Quay into one.

Maybe we encrypt it before storing?

[1] https://datatracker.ietf.org/doc/html/rfc7617
[2] https://datatracker.ietf.org/doc/html/rfc7235#section-4.1
[3] https://datatracker.ietf.org/doc/html/rfc7235#section-3.2
[4] https://datatracker.ietf.org/doc/html/rfc7235#section-4.3
[5] https://datatracker.ietf.org/doc/html/rfc7235#section-4.4

Reply:
I wasn't talking about basic auth. Looks like you're trying to save the result of the exchange flow in Redis.
Couldn't you do something like append it to the Quay one back in the response? That way, both are available, either for auth against Quay or the upstream registry, depending on the namespace type, and you don't have to store anything.

Reply:
Credentials for the upstream registry and tokens from the upstream registry should be kept secret and shouldn't be passed to Quay clients.

Reply:
FTR - we discussed on slack and decided the best way to go is to indeed keep the token in the backend. It was suggested there that we renew the token before it expires in the upstream registry, which is now done.